### PR TITLE
A preliminary implementation of visicode extension.

### PR DESCRIPTION
**Pull request by LiuHanzhang**
Only support image url for now.

### CODE REVIEW

**Comment by LiuHanzhang** [comment]
Only support image url for now.

File: visicode/example.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
+//!<visicode>
```
Comment (alanrtao) [inline visicode/example.cpp]: make an examples folder

File: visicode/src/test/suite/extension.test.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import * as assert from 'assert';
+
+// You can import and use all API from the 'vscode' module
+// as well as import your extension to test it
+import * as vscode from 'vscode';
+// import * as myExtension from '../../extension';
+
+suite('Extension Test Suite', () => {
```
Comment (alanrtao) [inline visicode/src/test/suite/extension.test.ts:8]: not really needed?

File: visicode/example.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  return ss.str();
+}
+
+} // namespace delegate
+} // namespace xnnpack
+} // namespace jit
+} // namespace torch
+
+//!<visicode>
+// https://media.npr.org/assets/img/2020/03/27/dualipareal-9f1d672307650164bada30d68c061cbf96777ee2.png
```
Comment (alanrtao) [inline visicode/example.cpp]: test with data url instead

File: visicode/src/extension.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	const scriptUri = panel.webview.asWebviewUri(scriptPath)
+
+	let code: string = editor.document.getText()
+	let html = `<link rel="stylesheet" href="${cssUri}">\n<script src="${scriptUri}"></script>\n\n`
+	
+	let startCodePos = 0
+	let startTagPos = 0
+	let endTagPos = 0
+	while (true) {
+		startTagPos = code.indexOf("//!<visicode>", startTagPos)
```
Comment (alanrtao) [inline visicode/src/extension.ts:58]: <visicode> should be enough? not sure what the //! does

File: visicode/src/extension.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		if (startTagPos == -1)
+			break
+		endTagPos = code.indexOf("//!</visicode>", endTagPos)
```
Comment (alanrtao) [inline visicode/src/extension.ts:61]: same with prev comment

File: visicode/src/extension.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	const cssUri = panel.webview.asWebviewUri(cssPath)
+	const scriptUri = panel.webview.asWebviewUri(scriptPath)
+
+	let code: string = editor.document.getText()
+	let html = `<link rel="stylesheet" href="${cssUri}">\n<script src="${scriptUri}"></script>\n\n`
+	
+	let startCodePos = 0
+	let startTagPos = 0
+	let endTagPos = 0
+	while (true) {
```
Comment (alanrtao) [inline visicode/src/extension.ts:57]: can change to for loop, while true looks scary lol
```
for (let startTagPos = ..., endTagPos = ...; startTagPos != -1; startTagPos = ..., endTagPos = ...) {
...
}
```

File: visicode/src/extension.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+		let codeSnippet: string = code.substring(startCodePos, startTagPos)
+		let taggedCodeSnippet: string = hljs.highlightAuto(codeSnippet).value
+		html = 	html
+				+ `<pre><code class="language-typescript" style="font-family: 'Courier New', Courier, monospace; font-size: 18px;">\n`
+				+ taggedCodeSnippet
+				+ `\n</code></pre>\n\n`
+
+		let tag: string = code.substring(startTagPos + "//!<visicode>".length, endTagPos)
+		let url = tag.substring(tag.indexOf("//") + 2)
+		html = 	html + `<img src="${url}">\n\n`
```
Comment (alanrtao) [inline visicode/src/extension.ts]: https://stackoverflow.com/questions/25551088/how-to-show-image-using-data-url-in-html-page

The basic one is just getting from API. You can add caching (with another js library maybe) if downloading every time is bad.

Alternatively if the API returns a markdown blob just use https://www.npmjs.com/package/markdown-it or other library to parse it into HTML.

**Review by alanrtao — CHANGES_REQUESTED** [review]

File: visicode/src/test/suite/extension.test.ts (hunk 2 above)
Comment (LiuHanzhang) [reply visicode/src/test/suite/extension.test.ts:8]: It's auto-generated by yo. We can delete it of course

File: visicode/src/extension.ts (hunk 4 above)
Comment (LiuHanzhang) [reply visicode/src/extension.ts:58]: You are right. Not all languages start comment with //

Comment (LiuHanzhang) [reply visicode/src/extension.ts:58]: It's a lot more difficult to consider all possible comment symbols. Just assume we use languages whose comments start with '//'. Perhaps we can handle this later

**Review by alanrtao — APPROVED** [review]

**Pull request merged**